src/misc/umap.c

```diff
@@ -0,0 +1,262 @@
+/* -*- mode: C -*-  */
+/*
+   IGraph library.
+   Copyright (C) 2008-2021  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+   */
+
+#include "igraph_matrix.h"
+#include "igraph_interface.h"
+#include "igraph_umap.h"
+#include "igraph_constructors.h"
+#include "igraph_layout.h"
+
+static igraph_error_t igraph_get_k_smallest(igraph_vector_t *values,
+        igraph_vector_int_t *smallest, igraph_integer_t k, igraph_integer_t starting_index)
+{
+    igraph_vector_t smallest_values;
+    igraph_real_t value;
+    igraph_integer_t max_index;
+
+    IGRAPH_VECTOR_INIT_FINALLY(&smallest_values, 0);
+    igraph_vector_int_resize(smallest, 0);
+    for (igraph_integer_t i = starting_index; i < igraph_vector_size(values); i++) {
+        value = VECTOR(*values)[i];
+        if (igraph_vector_int_size(smallest) < k) {
+            igraph_vector_int_push_back(smallest, i);
+            igraph_vector_push_back(&smallest_values, i);
+        } else {
+            max_index = 0;
+            for (igraph_integer_t j = 1; j < igraph_vector_int_size(smallest); j++) {
+                if (VECTOR(smallest_values)[j] > VECTOR(smallest_values)[max_index]) {
+                    max_index = j;
+                }
+            }
+            if (value > VECTOR(smallest_values)[max_index]) {
+                VECTOR(smallest_values)[max_index] = value;
+                VECTOR(*smallest)[max_index] = i;
+            }
+        }
+    }
+    igraph_vector_destroy(&smallest_values);
+    IGRAPH_FINALLY_CLEAN(1);
+    return (IGRAPH_SUCCESS);
+}
+
+/*This creates a k-nearest neighbors graph with their distances.*/
```

Comment:
My intention would be to work with an *existing* graph (usually knn) instead of making one from whatever vector space, so that we can trash this part. Indeed, the perception that distance computation (e.g. correlation, Euclidean), graph construction (via knn or whatever other method), and layouting are one thing is the source of many confusions in the field I think.

Reply:
For now, whatever you find most useful, but the function signature at the end of the day would be expected... ehr, at least *I* would expect it - somewhat naively perhaps - to be:

graph pointer -> layout pointer

instead of:

data pointer -> layout pointer

Reply:
> My intention would be to work with an existing graph

I would like to ask to briefly document these things, i.e. what the intention is, in the top post. Hopefully it's not too much work—the briefest note will do. The idea is that those of is who can't follow comments daily can still catch up later, without having to read through the entire thread, and go through past ideas that already got discarded during the journey.